Improve error message if data dir and config unavailable due to permissions

stevenGravy opened the pull request:
addresses #10032 where user gets very cryptic error message if running w/o rights to data dir.

If a user starts `teleport` and the process is running without rights to create the data dir or maintain file in data dir will now get this message.  

```code
ERRO             Teleport process must have rights to maintain files in data directory: /var/lib/teleport. Run as root or with sudo if not running as process that can maintain that directory.  service/service.go:664
```

stevenGravy committed:
Provider error info on data dir rights

russjones commented on lib/service/service.go:
```diff
@@ -621,8 +621,10 @@ func NewTeleport(cfg *Config) (*TeleportProcess, error) {
 	// create the data directory if it's missing
 	_, err = os.Stat(cfg.DataDir)
 	if os.IsNotExist(err) {
+		cfg.Log.Errorf("Teleport process must have rights to maintain data directory: %v. Run as root or with sudo if not running as process that can maintain that directory. ", cfg.DataDir)
```
@r0mant @zmb3 Can you guys think of better error messsage?

stevenGravy replied:
@zmb3 thank you.  I have corrected the location to be after doing the mkdir and tested again.

zmb3 replied:
Okay, we've got the right location now, but I still think we need to work out a better message.

As a security focused company I'm hesitant to say "just run as root if your permissions aren't correct"

stevenGravy replied:
sorry, resolved before continuing @zmb3 's point.  Should we mention `sudo`, `root` is required just for `ssh` and services.

stevenGravy replied:
Circling back on this one.  @xinding33 @zmb3 @russjones any suggestions for a error message?

stevenGravy committed:
Correct location of error message

zmb3 commented:
We really should be confirming that this is actually a permission error before we log this error. Something like: 
```
if errors.Is(err, fs.ErrPermission) {
    ...
```

As for the error message, how about something as simple as:

`Teleport does not have permission to write to <dir>. Ensure that you are running as a user with appropriate permissions.`

stevenGravy committed:
Modified message and added in comparsion that it is a permission error

stevenGravy committed:
Added similar message for appropiate access when trying to use a Teleport configuration file (/etc/teleport.yaml) and it fails to load due to permission error.

stevenGravy committed:
lint fixes

stevenGravy commented:
> 



> We really should be confirming that this is actually a permission error before we log this error. Something like:
> 
> ```
> if errors.Is(err, fs.ErrPermission) {
>     ...
> ```
> 
> As for the error message, how about something as simple as:
> 
> `Teleport does not have permission to write to <dir>. Ensure that you are running as a user with appropriate permissions.`

Thanks @zmb3 I have set this as the output and permission error handling.  I also modified so that if the user can't access the `/etc/teleport.yaml` they get more context on what the file is for.  It also only gives the permissions message after checking the permission error match.

Before:
```bash
ERROR: failed to open file: /etc/teleport.yamlold
open /etc/teleport.yamlold: permission denied
```

Updated to:
```bash
ERROR: failed to open file for Teleport configuration: /etc/teleport.yamlold. Ensure that you are running as a user with appropriate permissions. 
```

zmb3 commented on lib/config/fileconf.go:
```diff
@@ -91,7 +93,10 @@ type FileConfig struct {
 	if err != nil {
-		return nil, trace.Wrap(err, fmt.Sprintf("failed to open file: %v", filePath))
+		if errors.Is(err, fs.ErrPermission) {
+			return nil, trace.Wrap(err, fmt.Sprintf("failed to open file for Teleport configuration: %v. Ensure that you are running as a user with appropriate permissions.", filePath))
```
```suggestion
			return nil, trace.Wrap(err, "failed to open file for Teleport configuration: %v. Ensure that you are running as a user with appropriate permissions.", filePath)
```

No need for the `Sprintf`. (same comment applies below as well)

zmb3 commented on lib/utils/utils.go:
```diff
@@ -439,6 +441,10 @@ func ReadHostUUID(dataDir string) (string, error) {
 	err := ioutil.WriteFile(filepath.Join(dataDir, HostUUIDFile), []byte(id), os.ModeExclusive|0400)
 	if err != nil {
+		if errors.Is(err, fs.ErrPermission) {
+			//do not convert to system error as this loses the ability to compare that it is a permission error
```
`trace.ConvertSystemError` will convert it to an access denied error, so another option would be to leave this as-is, and update call-site to check `if trace.IsAccessDenied(err) { ...`

stevenGravy replied:
Leaving as is after discussing with @zmb3 , will likely revisit in the future.

zmb3 reviewed (commented):
Just a couple minor comments.

stevenGravy committed:
Update lib/config/fileconf.go

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

zmb3 commented on lib/utils/utils.go:
```diff
@@ -439,6 +441,10 @@ func ReadHostUUID(dataDir string) (string, error) {
 func WriteHostUUID(dataDir string, id string) error {
 	err := ioutil.WriteFile(filepath.Join(dataDir, HostUUIDFile), []byte(id), os.ModeExclusive|0400)
```
```suggestion
	err := os(filepath.Join(dataDir, HostUUIDFile), []byte(id), os.ModeExclusive|0400)
```

Unrelated, but while you're here..

stevenGravy committed:
Update lib/utils/utils.go

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

zmb3 commented on lib/config/fileconf.go:
```diff
@@ -91,7 +93,10 @@ type FileConfig struct {
+		if errors.Is(err, fs.ErrPermission) {
+			return nil, trace.Wrap(err, "failed to open file for Teleport configuration: %v. Ensure that you are running as a user with appropriate permissions.", filePath)
+		}
+		return nil, trace.Wrap(err, fmt.Sprintf("failed to open file for Teleport configuration: %v.", filePath))
```
```suggestion
		return nil, trace.Wrap(err, "failed to open Teleport configuration at %v", filePath)
```

zmb3 commented on lib/utils/utils.go:
```diff
@@ -437,8 +439,12 @@ func ReadHostUUID(dataDir string) (string, error) {
 // WriteHostUUID writes host UUID into a file
 func WriteHostUUID(dataDir string, id string) error {
-	err := ioutil.WriteFile(filepath.Join(dataDir, HostUUIDFile), []byte(id), os.ModeExclusive|0400)
+	err := os(filepath.Join(dataDir, HostUUIDFile), []byte(id), os.ModeExclusive|0400)
```
My original suggestion was invalid. You want `os.WriteFile` instead of `ioutil.WriteFile`. I mistakenly suggested `os` without the `.WriteFile` part.

stevenGravy replied:
fixing, thanks!

stevenGravy committed:
modify write file call

stevenGravy committed:
Update lib/config/fileconf.go

Co-authored-by: Zac Bergquist <zmb3@users.noreply.github.com>

stevenGravy committed:
update read from file message so test passes

stevenGravy committed:
Merge branch 'master' into stevenGravy/starterrormessage

zmb3 committed:
Merge branch 'master' into stevenGravy/starterrormessage

stevenGravy committed:
Modifications to allow for detecting file error permissions and
put out helpful error messages.

stevenGravy committed:
Merge branch 'master' into stevenGravy/starterrormessage

stevenGravy committed:
lint fix

stevenGravy committed:
lint fixes

stevenGravy committed:
Merge branch 'master' into stevenGravy/starterrormessage

stevenGravy committed:
lint fix

stevenGravy committed:
Merge branch 'stevenGravy/starterrormessage' of https://github.com/gravitational/teleport into stevenGravy/starterrormessage
update

stevenGravy committed:
lint fix
